[BOOST-562] read models documentation (#264)
boostercloud/booster: docs/README.md

Despite you can place your read models in any directory, we strongly recommend you to put them in `<project-root>/src/read-models`. Having all the read models in one place will help you to understand your application's capabilities at a glance.

```text
<project-root>
├── src
│   ├── commands 
│   ├── common
│   ├── config
│   ├── entities
│   ├── read-models  <------ put them here
│   ├── events
│   ├── index.ts
│   └── read-models
```

#### Creating a read model

The preferred way to create a read model is by using the generator, e.g.

```shell
boost new:read-model CartReadModel --fields id:UUID cartItems:"Array<CartItem>" paid:boolean --projects Cart
```

The generator will create a Typescript class under the read-models directory `<project-root>/src/read-models/CartReadModel.ts`.

Read Model classes can also be created by hand and there are no restrictions. The structure of the data is totally open and can be as complex as you can manage in your projection functions.

#### The projection function
A `Projection` is a method decorated with the `@Projects` decorator that, given a new entity value and (optionally) the current read model state, generate a new read model value.

> Read models can be projected from multiple [entities](#4-entities-and-reducers) as soon as they share some common key called `joinKey`. The `joinKey` is an entity field used to join data for projections. Read models can be considered as reactive join operations that can listen to different entities that has common fields between them. Let's see an example:

Reviewer: ```suggestion
Read models can be projected from one or more [entities](#4-entities-and-reducers) as soon as all the entities involved have one field with the same semantics that can be used as a join key. A join key in Booster is similar to join keys in relational databases, they're a field that contains a value that can be used to relate one entity with other (usually an identifier or a reference to other entity). When an entity is updated, Booster uses the join key to find the right read model instance, so all entities that share the same join key value will trigger the projections of the same read model instance. When defining a projection with the `@Projects` decorator, it's required to set the field name of the join key in each entity. Let's see an example:
```

```typescript
@ReadModel
export class UserReadModel {
  public constructor(readonly username: string, /* ...(other interesting fields from users)... */) {}
  
  @Projects(User, 'id')
  public static projectUser(entity: User, current?: UserReadModel) { // Here we update the user fields}

  @Projects(Post, 'ownerId')
  public static projectUserPost(entity: Post, current?: UserReadModel) { //Here we can adapt the read model to show specific user information related with the Post entity}
}
```
In the previous example we are projecting the `User` entity using the user `id` and also we are projecting the `User` entity based on the `ownerId` of the `Post` entity

#### Authorizing read models

> Read models are part of the public API of a Booster application, so you can define who is authorized to query them. The Booster authorization feature is covered in [this](#authentication-and-authorization) section. So far, we have seen that you can make a read model publicly accessible by authorizing `'all'` to query it. You can also set specific roles providing the users role array in this way: `authorize: [Admin]`.

Reviewer: ```suggestion
Read models are the tool to build the public read API of a Booster application, so you can define who is authorized to query and subscribe to them. The Booster authorization feature is covered in [the auth section](#authentication-and-authorization). So far, we have seen that you can make a read model publicly accessible by authorizing `'all'` to query it or you can set specific roles providing an array of roles in this way: `authorize: [Admin]`.
```

#### Querying a read model

> Read models are the read API for Booster applications and they can be accessed using the GraphQL queries. For every read model, Booster automatically creates all the necessary queries. For example, given this `CartReadModel` read model:

Reviewer: ```suggestion
For every read model, Booster automatically creates all the necessary queries and subscriptions. For example, given this `CartReadModel`:
```

```typescript
@ReadModel

> export class NameReadModel {

Reviewer: Y U NO...

```suggestion
export class ReadModelName {
```

> export class NameReadModel {

Author: agree!

  public constructor(
    readonly fieldA: SomeType,
    readonly fieldB: SomeType,
    /* as many fields as needed */
  ) {}

  @Projects(SomeEntity, 'entityField')
  public static projectionName(entity: SomeEntity, currentEntityReadModel?: NameReadModel): NameReadModel {
    return new NameReadModel(/* initialize here your constructor properties */)
  }
  
  @Projects(SomeEntity, 'othetEntityField')
  public static projectionName(entity: SomeEntity, currentEntityReadModel?: NameReadModel): NameReadModel {
    return new NameReadModel(/* initialize here your constructor properties */)
  }
  /* as many projections as needed */
}
```

#### Read models naming convention

> As it has been previously commented, but semantics plays an important role in designing a coherent system and your application should reflect your domain concepts. Altough you have freedom to choose your read model names, we strongly recommend you to name them using the name of the entity that they are projecting followed by `ReadModel` suffix. 

Author: agree!!

> Despite you can place commands, and other Booster files, in any directory, we strongly recommend you to put them in `<project-root>/src/read-models`. Having all the commands in one place will help you to understand your application's capabilities at a glance.

Author: fixed!

```text
<project-root>
├── src
│   ├── commands 
│   ├── common
│   ├── config
│   ├── entities
│   ├── read-models  <------ put them here
│   ├── events
│   ├── index.ts
│   └── read-models
```

#### Creating a read model

The preferred way to create a read model is by using the generator, e.g.

```shell
boost new:read-model CartReadModel --fields id:UUID cartItems:"Array<CartItem>" paid:boolean --projects Cart
```

The generator will create a Typescript class under the read-models directory `<project-root>/src/read-models/CartReadModel.ts`.

> Read Model classes can also be created by hand and there are no restrictions regarding the place you put the files. The structure of the data is totally open and can be as complex as you can manage in your projection functions.

Author: applied the change and removed any mention to name guidelines

#### The projection function
A `Projection` is a method decorated with the `@Projects` decorator that, given a new entity value and (optionally) the current read model state, generate a new read model value.

> Read models can be projected from multiple [entities](#4-entities-and-reducers) as soon as they share some common key called `joinKey`. The `joinKey` is the field used across entities to relaunch the projection function. To clarify this let's see an example:

Author: To be honest I don't have clear the concept of the `joinKey` to explain it properly

> Read models can be projected from multiple [entities](#4-entities-and-reducers) as soon as they share some common key called `joinKey`. The `joinKey` is the field used across entities to relaunch the projection function. To clarify this let's see an example:

Reviewer 2: @rdiaz82 Does this answer from Javier help you to understand it?

(look at the part where he answers: _"I wonder whose our most prolific user– I should totally reach out to them"_)

https://spectrum.chat/boostercloud/general/data-access~4c5eccde-3e40-4282-92d4-9d3342dc0bbf?m=MTU5NjA1Nzc4NzU5OQ==

If not, we can have a meeting so that you can understand it and finis this part of the explanation 😉

```typescript
@Projects(Cart, "id")
	public static projectCart(entity:Cart, currentReadModel: CartReadModel): CartReadModel {
		return new CartReadModel(entity.id, entity.items)

> 	}

Author: applied

```

> In the previous code, the `id` property is used as `joinKey`. This means that each time that a `Cart` that has the same `id` as the read model's `id` is modified, the projection is called.

Author: changed

#### Authorizing read models

> Read models are part of the public API of a Booster application, so you can define who is authorized to query them. The Booster authorization feature is covered in [this](https://github.com/boostercloud/booster/blob/master/docs/README.md#authentication-and-authorization) section. So far, we have seen that you can make a read command publicly accessible by authorizing `'all'` to query it. You can also set specific roles providing the users role array in this way: `authorize: [Admin]`.

Author: applied

#### Querying a read model

> Read models are the read API for Booster applications. they can be accessed using the GraphQL queries. When you create a read model Booster automatically creates all the neccesary queries for your read models, e.g., given this `CartReadModel` read model:

Author: applied

```typescript
@ReadModel({
	authorize: 'all'
})
export class CartReadModel {
	public constructor(
		public id: UUID,
		readonly items: Array<CartItem>,
	) {}

	@Projects(Cart, "id")
	public static projectCart(entity:Cart, currentReadModel: CartReadModel): CartReadModel {
		return new CartReadModel(entity.id, entity.items)
	}

> }

Author: applied!

```

> if you hit the provided Url at the end of the deploy process:

Author: applied

```
<httpURL>/graphql
```

and you also provide this query using any graphQL client:

```graphQL
query {
	CartReadModel(id: "012345") {
		id
		items
	}
}
```

you can get some results in the following format:

```json
{
	data: {
		"CartReadModel": {
			"id" : "012345",
			"items": [
				{
					"sku":"ABC_01",
					"quantity": 1
				}
			]
		}
	}
}

> ```

Reviewer 2: As this is constantly changing now, I would use the same approach we followed when explaining the same sections for commands, and only talk about one of the queries that are generated. I would not say anything about the URL, and the format of the response. For that, I just refer to the GraphQL API section.

In this case, the query that is generated is: 
```graphql
query CartReadModel(id: ID!): CartReadModel
```

Then I would just say something like "For more information about all the "queries" and how to use them, please go to section (GraphQL API section)"

Author: I made this section generic just applying @alvaroloes comment's

#### Getting real-time updates for a read model

> Booster GraphQL API also provides support for real-time updates using subscriptions and websocket, to get more information about it go to the [GraphQL API](https://github.com/boostercloud/booster/blob/master/docs/README.md#subscribing-to-read-models) section.

Reviewer 3: ```suggestion
Booster GraphQL API also provides support for real-time updates using subscriptions and websocket, to get more information about it go to the [GraphQL API](#subscribing-to-read-models) section.
```
